fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/AbstractDataAccessRule.java

```diff
@@ -218,13 +223,62 @@ protected Set<? extends RelationalExpression> dataAccessForMatchPartition(@Nonnu
+        // is actually useful in order to eagerly avoid creating useless plan variations. An intersection of
+        // n accesses, if it can be planned, produces an ordering which can be computed by either intersecting all
+        // n orderings of all n accesses, or by looking up the ordering of subpartitions of these n accesses, which
+        // are always planned before, and intersecting any such orderings that cover all n accesses. One observation of
+        // doing that is that a new intersection produces (in loose terms) the same ordering as any of its constituent
+        // subpartitions except for (potentially) additional equality-bound values. We attempt to optimize by
+        // disallowing an intersection where any subpartition of that intersection has the same equality-bound values
+        // that the new intersection would have as that intersection can be regarded as non-filtering or useless.
+        //
+        // Example:
+        //   I1: a, id
+        //   I2: b, id
+        //   query: SELECT * FROM T WHERE a = 5 and b = 'foo'
+        //   Intersection of I1 ∩ I2 is useful producing an intersection order on id with a, b equality-bound
+        //
+        //   If we add
+        //     I3: a, b, id
+        //   we consider it useless to intersect I1 or I2 or I1 ∩ I2 with it as all of these subpartitions already
+        //   constrain a and b.
+        //
+        // We further apply the same idea for an intersection we plan after the current intersection
+        // which is a super-partition of the current intersection, i.e. it intersects all the current accesses, but it
+        // also further intersects more accesses that also impose new useful equality-bound values (see above). Now,
+        // the current partition of accesses is useless as the one planned subsequently is better.
+        //
+        // Alternative view: Mathematically, we consider the power set of all accesses when we plan the intersections
+        // of accesses. Each item in that set is a partition. There is a partial order PO defined on the power set that
+        // states that two elements a, b: aPOb if the set of accesses in b includes the accesses in a.
+        // Some partitions cannot be used outright as they e.g. would not even range over compatibly-ordered data
+        // streams. These partitions are evicted from the power set. All partitions containing those partitions are
+        // immediately evicted as well. Apart from those partitions, there are other albeit useless/redundant elements
+        // (partitions) in the power; we evict those and the ones containing these useless partitions in the same
+        // fashion (as they are also useless/redundant).
+        // What we end up with is a subset of the power set of all accesses that are possible to intersect,
+        // that are not considered useless/redundant and that are local maximums with respect to PO.
```

Comment:
I think what you're describing is very similar to a Hasse diagram of power set, if we think of this hypercube as a directed graph where direction indicates set containment, then the algorithm works by identifying a node that can be deleted from the graph, and iteratively removing all of the nodes along all paths originating from that node.

Reply:
That's absolutely correct.